File: python/13_btree.py

"""
Binary Tree implementation in Python.
"""

from typing import Optional


class Node:
    """
    A node in a binary tree.
    """

    def __init__(self, key: int) -> None:
        """Initialize a tree node with value and left/right children."""
        self.left: Optional['Node'] = None
        self.right: Optional['Node'] = None
        self.val: int = key

    def traverse_preorder(self) -> None:
        """Preorder traversal (root, left, right)."""
        print(self.val, end=' ')
        if self.left:
            self.left.traverse_preorder()
        if self.right:
            self.right.traverse_preorder()

    def traverse_inorder(self) -> None:
        """Inorder traversal (left, root, right)."""
        if self.left:
            self.left.traverse_inorder()
        print(self.val, end=' ')
        if self.right:
            self.right.traverse_inorder()

    def traverse_postorder(self) -> None:
        """Postorder traversal (left, right, root)."""
        if self.left:
            self.left.traverse_postorder()
        if self.right:
            self.right.traverse_postorder()
        print(self.val, end=' ')


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    print("Pre order Traversal: ", end="")
    root.traverse_preorder()
    print("\nIn order Traversal: ", end="")
    root.traverse_inorder()
    print("\nPost order Traversal: ", end="")
    root.traverse_postorder()

File: java/13_btree.java
/**
 * Simple binary tree demonstration with recursive traversals (preorder,
 * inorder, postorder). The implementation preserves the traversal logic
 * used by the original example; the changes here are purely stylistic and
 * API additions for readability.
 */

// Node creation
class Node {
    int key;
    Node left, right;

    /**
     * Create a new tree node containing {@code item}.
     */
    public Node(int item) {
        key = item;
        left = right = null;
    }
}

class BinaryTree {
    Node root;

    /**
     * Construct a binary tree with a single root node containing {@code key}.
     */
    BinaryTree(int key) {
        root = new Node(key);
    }

    /**
     * Construct an empty binary tree.
     */
    BinaryTree() {
        root = null;
    }

    // Traverse Inorder (left, root, right)
    public void traverseInOrder(Node node) {
        if (node != null) {
            traverseInOrder(node.left);
            System.out.print(" " + node.key);
            traverseInOrder(node.right);
        }
    }

    // Convenience wrapper: traverse the whole tree inorder
    public void traverseInOrder() {
        traverseInOrder(root);
    }

    // Traverse Postorder (left, right, root)
    public void traversePostOrder(Node node) {
        if (node != null) {
            traversePostOrder(node.left);
            traversePostOrder(node.right);
            System.out.print(" " + node.key);
        }
    }

    // Convenience wrapper: traverse the whole tree postorder
    public void traversePostOrder() {
        traversePostOrder(root);
    }

    // Traverse Preorder (root, left, right)
    public void traversePreOrder(Node node) {
        if (node != null) {
            System.out.print(" " + node.key);
            traversePreOrder(node.left);
            traversePreOrder(node.right);
        }
    }

    // Convenience wrapper: traverse the whole tree preorder
    public void traversePreOrder() {
        traversePreOrder(root);
    }

    public static void main(String[] args) {
        BinaryTree tree = new BinaryTree();

        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);

        System.out.print("Pre order Traversal: ");
        tree.traversePreOrder();
        System.out.print("\nIn order Traversal: ");
        tree.traverseInOrder();
        System.out.print("\nPost order Traversal: ");
        tree.traversePostOrder();
    }
}